File: ensarkbank-android/app/src/main/java/com/ensark/ensarkbank/repository/CardRepository.java
package com.ensark.ensarkbank.repository;

import android.content.Context;

import com.ensark.ensarkbank.api.ApiClient;
import com.ensark.ensarkbank.api.CardApiService;
import com.ensark.ensarkbank.model.dto.CardRequest;
import com.ensark.ensarkbank.model.dto.CardResponse;
import com.ensark.ensarkbank.model.dto.CardUsageResponse;
import com.ensark.ensarkbank.model.dto.CardSettingsRequest;
import com.ensark.ensarkbank.model.dto.PinChangeRequest;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

import retrofit2.Callback;

public class CardRepository {

    private final CardApiService apiService;

    public CardRepository(Context context) {
        apiService = ApiClient.getClient(context).create(CardApiService.class);
    }

    public void apply(CardRequest request, Callback<CardResponse> callback) {
        apiService.apply(request).enqueue(callback);
    }

    public void findByCustomerEmail(String email, Callback<List<CardResponse>> callback) {
        apiService.findByCustomerEmail(email).enqueue(callback);
    }

    public void updatePin(Long id, PinChangeRequest request, Callback<CardResponse> callback) {
        apiService.updatePin(id, request).enqueue(callback);
    }

    public void findById(Long id, Callback<CardResponse> callback) {
        apiService.findById(id).enqueue(callback);
    }

    public void setTransactionLimit(Long id, BigDecimal dailyLimit, BigDecimal monthlyLimit, Callback<CardResponse> callback) {
        apiService.setTransactionLimit(id, dailyLimit, monthlyLimit).enqueue(callback);
    }

    public void reportLostOrStolen(Long id, String reason, Callback<CardResponse> callback) {
        apiService.reportLostOrStolen(id, reason).enqueue(callback);
    }

    public void getUsage(Long id, Callback<CardUsageResponse> callback) {
        apiService.getUsage(id).enqueue(callback);
    }

    public void createSettingsRequest(Map<String, Object> body, Callback<CardSettingsRequest> callback) {
        apiService.createSettingsRequest(body).enqueue(callback);
    }

    public void getMySettingsRequests(Long customerId, Callback<List<CardSettingsRequest>> callback) {
        apiService.getMySettingsRequests(customerId).enqueue(callback);
    }
}

File: ensark-flutter/lib/repositories/card_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';
import '../models/card/card_models.dart';
import '../providers/core_providers.dart';

part 'card_repository.g.dart';

class CardRepository {
  final Dio _dio;

  CardRepository(this._dio);

  Future<CardResponse> apply(CardRequest request) async {
    final response = await _dio.post('api/card/apply', data: request.toJson());
    return CardResponse.fromJson(response.data);
  }

  Future<List<CardResponse>> findByCustomerEmail(String email) async {
    final response = await _dio.get('api/card/customer/$email');
    return (response.data as List).map((e) => CardResponse.fromJson(e)).toList();
  }

  Future<CardResponse> updatePin(int id, PinChangeRequest request) async {
    final response = await _dio.patch('api/card/$id/pin', data: request.toJson());
    return CardResponse.fromJson(response.data);
  }

  Future<CardUsageResponse> getUsage(int id) async {
    final response = await _dio.get('api/card/$id/usage');
    return CardUsageResponse.fromJson(response.data);
  }

  Future<CardResponse> setTransactionLimit(int id, double dailyLimit, double monthlyLimit) async {
    final response = await _dio.patch(
      'api/card/$id/limit',
      queryParameters: {'dailyLimit': dailyLimit, 'monthlyLimit': monthlyLimit},
    );
    return CardResponse.fromJson(response.data);
  }

  Future<CardResponse> reportLostOrStolen(int id, String reason) async {
    final response = await _dio.post(
      'api/card/$id/report-lost-stolen',
      queryParameters: {'reason': reason},
    );
    return CardResponse.fromJson(response.data);
  }

  Future<CardSettingsRequest> createSettingsRequest(Map<String, dynamic> body) async {
    final response = await _dio.post('api/card-settings-requests/', data: body);
    return CardSettingsRequest.fromJson(response.data);
  }
}

@riverpod
CardRepository cardRepository(Ref ref) {
  return CardRepository(ref.watch(dioProvider));
}
